packages/flutter/lib/src/cupertino/sheet.dart

```diff
@@ -0,0 +1,349 @@
+            reverseCurve: Curves.fastEaseInToSlowEaseOut.flipped,
+          ).drive(offsetTween);
+
+    return SlideTransition(
+      position: positionAnimation,
+      child: child,
+    );
+  }
+
+  static Widget _coverSheetSecondaryTransition(Animation<double> secondaryAnimation, Widget? child) {
+    const Curve curve = Curves.linearToEaseOut;
+    const Curve reverseCurve = Curves.easeInToLinear;
+    final Animation<double> curvedAnimation = CurvedAnimation(
+      curve: curve,
+      reverseCurve: reverseCurve,
+      parent: secondaryAnimation
+    );
+
+    return SlideTransition(
+      position: curvedAnimation.drive(_kMidUpTween),
+      transformHitTests: false,
+      child: ScaleTransition(
+        scale: curvedAnimation.drive(_kScaleTween),
+        filterQuality: FilterQuality.medium,
+        alignment: Alignment.topCenter,
+        child: ClipRRect(
+          borderRadius: const BorderRadius.vertical(top: Radius.circular(12)),
+          child: child,
+        ),
+      ),
+    );
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return ColoredBox(
+      color: CupertinoColors.transparent,
+      child: _coverSheetSecondaryTransition(
+        secondaryRouteAnimation,
+        _coverSheetPrimaryTransition(
+          context,
+          primaryRouteAnimation,
+          ClipRRect(
+            borderRadius: const BorderRadius.vertical(top: Radius.circular(12)),
+            child: child,
+          ),
+        ),
+      ),
+    );
+  }
+}
+
+/// Docs placeholder
+class CupertinoSheetRoute<T> extends PageRoute<T> with CupertinoSheetRouteTransitionMixin<T> {
+  /// Creates a page route for use in an iOS designed app.
+  CupertinoSheetRoute({
+    required this.pageBuilder,
+  });
+
+  /// Docs placeholder
+  final WidgetBuilder pageBuilder;
+
+  @override
+  DelegatedTransitionBuilder? get delegatedTransition => CupertinoSheetTransition.delegateTransition;
+
+  @override
+  Widget buildContent(BuildContext context) {
+    final BuildContext? topLevelContext = CupertinoSheetController.maybeOf(context)?.topLevelContext;
+    return CupertinoSheetController(
+      context: context,
+      topLevelContext: topLevelContext ?? context,
+      child: pageBuilder(context),
+    );
+  }
+
+  /// Docs placeholder
+  static CupertinoSheetController? maybeOf(BuildContext context) {
+    return CupertinoSheetController.maybeOf(context);
+  }
+
+  /// Docs placeholder
+  static CupertinoSheetController of(BuildContext context) {
+    return CupertinoSheetController.of(context);
+  }
+
+  @override
+  Color? get barrierColor => const Color(0x20000000);
+
+  @override
+  bool get barrierDismissible => false;
+
+  @override
+  String? get barrierLabel => 'Stacked card appearance for modal bottom sheet';
+
+  @override
+  bool get maintainState => true;
+
+  @override
+  bool get opaque => false;
+}
+
+/// Docs placeholder
+class CupertinoSheetController extends InheritedWidget {
```

Comment:
This needs a better name. Controller is usually a changeNotifier object in Flutter.

Reply:
instead of a inherited widget that people can access different level of context, have you thought about have a controller object that provide API to pop or push navigator route at different level?

Reply:
The reason for this is that holding a context as widget parameter is can be error prompt as it can go stale at anytime. it will be more dangerous if this is a public api

Reply:
The only place that the `context` parameter is used is in popSheet, right? Maybe that could just take BuildContext as a parameter so that CupertinoSheetController doesn't need to.

Reply:
> instead of a inherited widget that people can access different level of context, have you thought about have a controller object that provide API to pop or push navigator route at different level?

@chunhtai Are you talking about a generic controller to control nested navigation, or something specifically for this modal bottom sheet?

Reply:
for this modal bottom sheet. For general nested navigation once can already use NavigatorState if they possess global key